sql/core/src/test/scala/org/apache/spark/sql/DataFrameComplexTypeSuite.scala

```diff
@@ -64,6 +68,24 @@ class DataFrameComplexTypeSuite extends QueryTest with SharedSparkSession {
     val ds100_5 = Seq(S100_5()).toDS()
     ds100_5.rdd.count
   }
+
+  test("SPARK-29503 nest unsafe struct inside safe array") {
+    withSQLConf(SQLConf.WHOLESTAGE_CODEGEN_ENABLED.key -> "false") {
+      val df = spark.sparkContext.parallelize(Seq(Seq(1, 2, 3))).toDF("items")
+
+      // items: Seq[Int] => items.map { item => Seq(Struct(item)) }
+      val result = df.select(
+        new Column(MapObjects(
+          (item: Expression) => array(struct(new Column(item))).expr,
```

Comment:
Hm, while fix seems fine to me too, was this only the reproducer? `MapObjects` is supposed to be internal purpose - it's under catalyst package.

Reply:
I haven't spent another time to try it (as it seems to be clean and simple reproducer). Btw, I'm not sure it's not going to be valid reproducer since it pulls catalyst package. Catalyst could analyze the query and inject it if necessary in any way.

I indicated you'd like to revisit #25745 - that was WIP and it didn't have any number of performance gain. I'd rather choose "safeness" over "speed", and even we haven't figured out there's outstanding difference between twos. It was the only one case MapObjects could have unsafe struct, by allowing this, safe and unsafe are possibly mixed up leading to encounter corner case.

Reply:
Yeah, I am not against this change. In that way, I think this fix is fine but wanted to know if this actually affects any user-facing surface.

Was wondering if we can benefit from #25745 since some investigations look already made there to completely use unsafe one instead.

Reply:
My guess of #25745 is, it was based on the assumption that it's safe to replace CreateNamedStruct with CreateNamedStructUnsafe as we already have one path to do this - and this observation broke the assumption. IMHO, once we found it's not safe to do this, the improvement has to prove safety before we take its benefits into account.